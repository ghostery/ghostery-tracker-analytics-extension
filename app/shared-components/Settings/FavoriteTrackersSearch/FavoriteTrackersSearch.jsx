import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';
import InputGroup from 'react-bootstrap/InputGroup';
import FormControl from 'react-bootstrap/FormControl';

import './FavoriteTrackersSearch.scss';

class FavoriteTrackersSearch extends Component {
  constructor(props) {
    super(props);
    this.state = { focusedResult: '' };
  }

  handleChange = (e) => {
    const { messageCreators } = this.props;
    messageCreators.searchBugDb(e.target.value);
    this.setState({ focusedResult: '' });
  }

  handleKeyDown = (e) => {
    e.stopPropagation();
    const { focusedResult } = this.state;
    const { bugDbSearchResults, panel } = this.props;
    if (e.target.value === '') { return; }
    if (bugDbSearchResults.length === 0) {
      this.setState({ focusedResult: '' });
      return;
    }

    // TODO: If we ever build for IE/Edge, we'll need to use e.keyCode instead
    // of e.key and ensure that the various key codes associated with each key
    // across different browsers are included
    switch (e.key) {
      case 'ArrowDown': {
        if (focusedResult === '') {
          this.setState({ focusedResult: `searchResult_${bugDbSearchResults[0].key}` });
          break;
        }

        const nextSibling = panel
          ? document.querySelector('.resolved').shadowRoot.getElementById(focusedResult).nextElementSibling
          : document.getElementById(focusedResult).nextElementSibling;

        if (nextSibling) {
          const nextBbox = nextSibling.getBoundingClientRect();
          const containerBbox = nextSibling.parentElement.getBoundingClientRect();
          if (nextBbox.bottom > containerBbox.bottom) {
            nextSibling.parentElement.scrollTop += 20;
          }
          this.setState({
            focusedResult: nextSibling.id,
          });
        }

        break;
      }

      case 'ArrowUp': {
        if (focusedResult === '') { break; }

        const previousSibling = panel
          ? document.querySelector('.resolved').shadowRoot.getElementById(focusedResult).previousElementSibling
          : document.getElementById(focusedResult).previousElementSibling;

        if (previousSibling) {
          const previousBbox = previousSibling.getBoundingClientRect();
          const containerBbox = previousSibling.parentElement.getBoundingClientRect();
          if (previousBbox.top < containerBbox.top) {
            previousSibling.parentElement.scrollTop -= 20;
          }
          this.setState({ focusedResult: previousSibling.id });
        } else {
          this.setState({ focusedResult: '' });
        }

        break;
      }

      case 'Enter': {
        let focusedOrSingleResult;
        if (bugDbSearchResults.length === 1) {
          focusedOrSingleResult = `searchResult_${bugDbSearchResults[0].key}`;
        } else {
          focusedOrSingleResult = focusedResult;
        }

        const focusedElement = panel
          ? document.querySelector('.resolved').shadowRoot.getElementById(focusedOrSingleResult)
          : document.getElementById(focusedOrSingleResult);
        const trackerName = focusedElement.getAttribute('trackername');
        const trackerId = focusedElement.getAttribute('trackerid');
        this.starOrRemoveTracker(trackerId, trackerName);
        this.setState({ focusedResult: '' });

        break;
      }
      default:
    }
  }

  starOrRemoveTracker = (trackerId, trackerName) => {
    const { actions, starredTrackerIds, messageCreators, panel, showToasts } = this.props;
    if (trackerId in starredTrackerIds) {
      if (showToasts) {
        actions.openToast({
          toastText: `${trackerName} is already a favorite tracker`,
        }, 2000);
      }
    } else {
      messageCreators.addStarredTracker({ trackerId, trackerName, view: panel ? '2' : '3' });
      if (showToasts) {
        actions.openToast({
          toastText: `${trackerName} has been added to favorite trackers`,
        }, 2000);
      }
    }
    this.searchRef.value = '';
    messageCreators.searchBugDb('');
  }

  handleRemoveStarred = (trackerId, trackerName) => {
    const { messageCreators, panel } = this.props;
    messageCreators.removeStarredTracker({
      trackerId, trackerName, view: panel ? '2' : '3',
    });
  }

  render() {
    const { bugDbSearchResults, starredTrackerIds, panel } = this.props;
    const { focusedResult } = this.state;

    const starredTrackerKeys = Object.keys(starredTrackerIds);
    starredTrackerKeys.sort((a, b) => (
      starredTrackerIds[a].toLowerCase().localeCompare(starredTrackerIds[b].toLowerCase())
    ));

    const starredListClasses = ClassNames('FavoriteTrackersSearch__starredList', {
      show: starredTrackerKeys.length > 0,
      panel,
    });

    return (
      <div className="FavoriteTrackersSearch__favoriteSettingsTabContent">
        <div className="FavoriteTrackersSearch__inputDropdownGroup">
          <div className="FavoriteTrackersSearch__favoriteTabHeader d-flex align-items-center justify-content-start">
            <img
              src={[chrome.extension.getURL('dist/images/shared/star-settings.svg')]}
              className="FavoriteTrackersSearch__starIcon"
              alt="Star icon"
            />
            <p>Your Favorite Trackers</p>
          </div>
          <InputGroup className="FavoriteTrackersSearch__searchInputContainer">
            <FormControl
              className="FavoriteTrackersSearch__searchInput"
              placeholder="Search"
              onChange={this.handleChange}
              onKeyDown={this.handleKeyDown}
              ref={(node) => { this.searchRef = node; }}
            />
          </InputGroup>
          <div className="FavoriteTrackersSearch__searchDropdown" style={{ display: `${bugDbSearchResults.length > 0 ? 'block' : 'none'}` }}>
            {bugDbSearchResults.map(tracker => (
              <div
                id={`searchResult_${tracker.key}`}
                key={tracker.key}
                trackername={tracker.name}
                trackerid={tracker.key}
                className={`FavoriteTrackersSearch__searchResult ${focusedResult === `searchResult_${tracker.key}` ? 'focused' : ''}`}
                onClick={() => this.starOrRemoveTracker(tracker.key, tracker.name)}
              >
                {tracker.name}
              </div>
            ))}
          </div>
          <p className="FavoriteTrackersSearch__favoriteText">
            Favorited trackers appear at the top of your list when they are detected on a page.
          </p>
        </div>
        <div className={starredListClasses}>
          {starredTrackerKeys.map(key => (
            <div className="FavoriteTrackersSearch__starredTracker  d-flex justify-content-between align-items-center" key={key}>
              {starredTrackerIds[key]}
              <img
                src={[chrome.extension.getURL('dist/images/shared/cross.svg')]}
                alt="Remove tracker from favorite list"
                onClick={() => this.handleRemoveStarred(key, starredTrackerIds[key])}
              />
            </div>
          ))}
        </div>
      </div>
    );
  }
}

FavoriteTrackersSearch.propTypes = {
  messageCreators: PropTypes.shape({
    searchBugDb: PropTypes.func.isRequired,
    removeStarredTracker: PropTypes.func.isRequired,
    addStarredTracker: PropTypes.func.isRequired,
  }).isRequired,
  showToasts: PropTypes.bool.isRequired,
  actions: PropTypes.shape({
    openToast: PropTypes.func.isRequired,
  }).isRequired,
  starredTrackerIds: PropTypes.objectOf(PropTypes.string).isRequired,
  bugDbSearchResults: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string,
      trackerName: PropTypes.string,
      trackerId: PropTypes.string,
    }),
  ).isRequired,
  panel: PropTypes.bool,
};

FavoriteTrackersSearch.defaultProps = {
  panel: false,
};

export default FavoriteTrackersSearch;
