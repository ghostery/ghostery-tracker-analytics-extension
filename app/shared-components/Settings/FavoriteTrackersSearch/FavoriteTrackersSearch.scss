@import '../../../utils/scss/color-variables.scss';
@import '../Settings.scss';

$header-height: 199px;

.FavoriteTrackersSearch__favoriteSettingsTabContent {
  width: 100%;
  height: inherit;
  display: flex;
  flex-direction: column;
}

.FavoriteTrackersSearch__favoriteTabHeader {
  & > img {
    margin-top: 15px;
    width: 13px;
    height: 13px;
  }

  & > p {
    margin: 15px 0 0 8px;
    font-weight: bold;
    font-size: 13px;
    color: $dark-grey;
  }
}

.FavoriteTrackersSearch__inputDropdownGroup {
  position: relative;
  margin: 0 auto;
  width: 87%;
}

.FavoriteTrackersSearch__searchInputContainer {
  width: 100%;
}

.FavoriteTrackersSearch__searchInput {
  margin-top: 11px;
  box-shadow: inset 0 1px 2px -.4px $opaque-background;
  box-sizing: border-box;
  outline: 0;
  border: 0;
  border-radius: 2.5px;
  padding-left: 10px;
  width: 100%;
  height: 25px;
  font-size: 11px;
  color: $dark-grey;
}

.FavoriteTrackersSearch__searchDropdown {
  position: absolute;
  background-color: $light-grey;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 3px 0;
  width: 100%;
  max-height: 200px;
  cursor: pointer;
}

.FavoriteTrackersSearch__searchResult {
  margin-left: 10px;
  width: 100%;
  height: 20px;
  font-size: 12px;
  line-height: 20px;
  text-align: left;

  &.focused,
  &:hover {
    color: $ghosty-blue;
  }
}

.FavoriteTrackersSearch__favoriteText {
  margin: 9px 0 20px 2px;
  width: 100%;
  font-weight: 600;
  font-size: 12px;
  text-align: left;
  color: $dark-grey;
}

.FavoriteTrackersSearch__starredList {
  margin: 10px 0 0;
  border-top: solid 1px $dark-grey;
  overflow-y: auto;
  min-height: 230px;
  max-height: 230px;
  display: none;

  &.show {
    display: block;
  }

  &.panel {
    max-height: calc(100vh - #{$header-height} - #{$footer-height});
  }
}

.FavoriteTrackersSearch__starredTracker {
  box-sizing: border-box;
  border-color: $dark-grey transparent transparent;
  border-width: 1px;
  border-style: solid;
  padding-left: 13px;
  width: 100%;
  height: 39px;
  font-weight: 600;
  font-size: 11px;
  line-height: 39px;
  text-align: left;
  color: $dark-grey;

  &:first-of-type {
    border-color: transparent;
  }

  &:last-of-type {
    border-color: $dark-grey transparent;
  }

  &:only-of-type {
    border-color: transparent transparent $dark-grey;
  }

  > img {
    margin-right: 15px;
    width: 10px;
    height: 10px;
    display: none;
    cursor: pointer;
  }

  &:hover {
    border-color: $ghosty-blue;
    color: $ghosty-blue;

    > img {
      display: block;
    }

    + .FavoriteTrackersSearch__starredTracker {
      border-color: transparent;
    }

    + .FavoriteTrackersSearch__starredTracker:last-of-type {
      border-color: transparent transparent $dark-grey;
    }
  }
}
