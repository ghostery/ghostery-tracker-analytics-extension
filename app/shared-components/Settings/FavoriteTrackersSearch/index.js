import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import FavoriteTrackersSearch from './FavoriteTrackersSearch.jsx';
import { openToast } from '../../../store/actions/UserInterfaceActions';
import {
  searchBugDb, addStarredTracker, removeStarredTracker,
} from '../../../store/messages/messageCreators';

const mapStateToProps = state => ({
  bugDbSearchResults: state.UserInterfaceReducer.bugDbSearchResults,
  starredTrackerIds: state.SettingsReducer.localSettings.starredTrackerIds,
  showToasts: state.SettingsReducer.localSettings.showToasts,
  messageCreators: { searchBugDb, addStarredTracker, removeStarredTracker },
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ openToast }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteTrackersSearch);
