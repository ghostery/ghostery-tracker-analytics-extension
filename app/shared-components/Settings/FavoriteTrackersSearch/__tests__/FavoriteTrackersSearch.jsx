import 'jsdom-global/register';
import React from 'react';
import { shallow, mount } from 'enzyme';
import chrome from 'sinon-chrome';

import FavoriteTrackersSearch from '../FavoriteTrackersSearch';

describe('app/shared-components/FavoriteTrackersSearch/', () => {
  describe('Snapshot tests with shallow', () => {
    global.chrome = chrome;
    test('FavoriteTrackersSearch component should render', () => {
      const data = {
        messageCreators: {
          searchBugDb: () => {},
          removeStarredTracker: () => {},
          addStarredTracker: () => {},
        },
        showToasts: true,
        actions: {
          openToast: () => {},
        },
        starredTrackerIds: { testing: 'starred' },
        bugDbSearchResults: [{ key: '78783743', trackerName: 'test', trackerId: '22' }],
        panel: true,
      };

      const component = shallow(<FavoriteTrackersSearch {...data} />);
      expect(component).toMatchSnapshot();
    });
  });
  describe('Functionality testing', () => {
    test('handleChange should resolve with the setState', () => {
      const data = {
        messageCreators: {
          searchBugDb: () => {},
          removeStarredTracker: () => {},
          addStarredTracker: () => {},
        },
        showToasts: true,
        actions: {
          openToast: () => {},
        },
        starredTrackerIds: {},
        bugDbSearchResults: [],
        panel: true,
      };

      const component = shallow(<FavoriteTrackersSearch {...data} />);
      component.setState({ focusedResult: '' });

      expect(component.find('.FavoriteTrackersSearch__searchInput').length).toBe(1);
    });

    test('handleRemoveStarred should be called on click', () => {
      const data = {
        messageCreators: {
          searchBugDb: () => {},
          removeStarredTracker: () => {},
          addStarredTracker: () => {},
        },
        showToasts: true,
        actions: {
          openToast: () => {},
        },
        starredTrackerIds: { testing: 'starred' },
        bugDbSearchResults: [{ key: '78783743', trackerName: 'test', trackerId: '22' }],
        panel: true,
      };

      const component = mount(<FavoriteTrackersSearch {...data} />);
      const handleRemoveStarredSpy = jest.spyOn(component.instance(), 'handleRemoveStarred');
      component.instance().forceUpdate();

      component.find('div.FavoriteTrackersSearch__starredTracker img').simulate('click');

      expect(handleRemoveStarredSpy).toHaveBeenCalled();
    });
    test('starOrRemoveTracker should be called on click', () => {
      const data = {
        messageCreators: {
          searchBugDb: () => {},
          removeStarredTracker: () => {},
          addStarredTracker: () => {},
        },
        showToasts: true,
        actions: {
          openToast: () => {},
        },
        starredTrackerIds: { testing: 'starred' },
        bugDbSearchResults: [{ key: '78783743', trackerName: 'test', trackerId: '22' }],
        panel: true,
      };

      const component = mount(<FavoriteTrackersSearch {...data} />);
      const starOrRemoveTrackerSpy = jest.spyOn(component.instance(), 'starOrRemoveTracker');
      component.instance().forceUpdate();

      component.find('div.FavoriteTrackersSearch__searchResult').simulate('click');

      expect(starOrRemoveTrackerSpy).toHaveBeenCalled();
    });
  });
});
