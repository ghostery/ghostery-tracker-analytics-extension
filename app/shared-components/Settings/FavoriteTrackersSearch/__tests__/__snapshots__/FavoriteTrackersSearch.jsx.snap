// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/shared-components/FavoriteTrackersSearch/ Snapshot tests with shallow FavoriteTrackersSearch component should render 1`] = `
<div
  className="FavoriteTrackersSearch__favoriteSettingsTabContent"
>
  <div
    className="FavoriteTrackersSearch__inputDropdownGroup"
  >
    <div
      className="FavoriteTrackersSearch__favoriteTabHeader d-flex align-items-center justify-content-start"
    >
      <img
        alt="Star icon"
        className="FavoriteTrackersSearch__starIcon"
        src={
          Array [
            undefined,
          ]
        }
      />
      <p>
        Your Favorite Trackers
      </p>
    </div>
    <InputGroup
      className="FavoriteTrackersSearch__searchInputContainer"
    >
      <FormControl
        className="FavoriteTrackersSearch__searchInput"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Search"
      />
    </InputGroup>
    <div
      className="FavoriteTrackersSearch__searchDropdown"
      style={
        Object {
          "display": "block",
        }
      }
    >
      <div
        className="FavoriteTrackersSearch__searchResult "
        id="searchResult_78783743"
        key="78783743"
        onClick={[Function]}
        trackerid="78783743"
      />
    </div>
    <p
      className="FavoriteTrackersSearch__favoriteText"
    >
      Favorited trackers appear at the top of your list when they are detected on a page.
    </p>
  </div>
  <div
    className="FavoriteTrackersSearch__starredList show panel"
  >
    <div
      className="FavoriteTrackersSearch__starredTracker  d-flex justify-content-between align-items-center"
      key="testing"
    >
      starred
      <img
        alt="Remove tracker from favorite list"
        onClick={[Function]}
        src={
          Array [
            undefined,
          ]
        }
      />
    </div>
  </div>
</div>
`;
